File: Data Structures & Algorithms/search-2d-matrix/submission-2.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        int l = 0;
        int r = m * n - 1;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            int row = mid / n;
            int col = mid % n;
            // cout << "row " << row << " col " << col << endl; 
            // cout << matrix[row][col] << endl;
            if (matrix[row][col] == target) return true;
            if (matrix[row][col] > target) r = mid - 1;
            else l = mid + 1;
        }
        return false;
    }
};

File: Data Structures & Algorithms/search-for-word/submission-2.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])
        vis = set()
        d = [(0,1),(1,0),(-1,0),(0,-1)]
        def bt(r, c, i, vis):
            if i == len(word):
                return True
            for dr, dc in d:
                row = r + dr
                col = c + dc
                if (row in range(m)
                and col in range(n)
                and (row, col) not in vis
                and word[i] == board[row][col]):
                    vis.add((row, col))
                    if bt(row, col, i+1, vis):
                        return True
                    vis.remove((row, col))
            return False
        for i in range(m):
            for j in range(n):
                if board[i][j] == word[0]:
                    vis.add((i, j))
                    if bt(i, j, 1, vis):
                        return True
                    vis.remove((i, j))
        return False